src/r_util.cpp

```diff
@@ -504,6 +506,37 @@ namespace rhost {
             return R_FalseValue;
         }
 
+        extern "C" SEXP save_project(SEXP id, SEXP project_name, SEXP dest_dir, SEXP temp_dir) {
+            auto blob_id = static_cast<blobs::blob_id>(Rf_asReal(id));
+            fs::path prj_name, d_dir, t_dir;
+
+            r_top_level_exec([&]() {
+                prj_name = fs::u8path(Rf_translateCharUTF8(STRING_ELT(project_name, 0)));
+                d_dir = fs::u8path(Rf_translateCharUTF8(STRING_ELT(dest_dir, 0)));
+                t_dir = fs::u8path(Rf_translateCharUTF8(STRING_ELT(temp_dir, 0)));
+            }, __FUNCTION__);
+
+            fs::path zip_file = t_dir / prj_name;
+            zip_file.replace_extension(".zip");
+
+            if (fs::exists(zip_file)) {
+                fs::remove(zip_file);
+            }
+
+            blobs::save_to_file(blob_id, zip_file);
+
+            fs::path temp_proj_dir = t_dir / prj_name;
+            if (fs::exists(temp_proj_dir)) {
+                fs::remove_all(temp_proj_dir);
+            }
+
+            fs::path dest_proj_dir = d_dir / prj_name;
+
+            rproj::extract_project(zip_file, dest_proj_dir, temp_proj_dir);
```

Comment:
Still not quite right. Same problem - there's a bunch of local stack-allocated variables here that have destructors that do meaningful cleanup (e.g. fs::path).

What you need is _all_ code that uses destructors to be inside extract_project, so that it can throw and unwind normally. Then wrap the whole enchilada with exceptions_to_errors. The function that calls exceptions_to_errors should not have any locals with non-trivial destructors.

---
In reply to: [74672717](https://github.com/Microsoft/R-Host/pull/78#discussion_r74672717) [](ancestors = 74672717)